### PR TITLE
Bump tuple inference length cutoff from 16 to 32

### PR DESCRIPTION
**Pull request by andyferris**
* ~~Bumps `tupletype_len` to `31`~~ Removes `tupletype_len` entirely.
 * Bumps `tuple_splat` to `32`
 * ~~Similarly for the types `Any16` -> `Any32` and `All16` -> `All32`.~~

For context, I feel this would be useful for working with heterogenously typed data as tuples and named tuples. In particular, for v0.7/v1.0, simple containers such as `Vector{NamedTuple{...}}` could be versatile, performant containers for tables and data (similarly for named tuples of arrays, etc), and at times the existing limits (where practically speaking its `14` elements being the biggest size that gives "full run-time speed") felt a bit limiting (e.g. a table with 15-30 columns doesn't seem particularly unreasonable, though for very large numbers I admit that switching to a more dynamic data structure might be preferable).

Incidentally, this might help with things like arrays with 15+ dimensions and so-on (#20163) (cc @Jutho). Having `30` dimensions as the maximum with fully-inferred code seems a somewhat reasonable cutoff number to me, giving ~1 billion elements for a 2x2x2x... sized array, as in #20163.

Of course, I'm more than a bit ignorant of what other impacts this may have internally for the compiler (compile time speed will obviously be slower in some situations) but I thought this might be worthwhile floating for inclusion in v0.7.

### CODE REVIEW

**Comment by andyferris** [comment]
Appveyer timed out, Travis is complaining about `Sockets`, and the other CIs passed.

We could also run nanosoldier, except it doesn't really measure the negative impacts (compile time) and I'm guessing the benchmark suite is likely to be insensitive to any positive impacts...

**Comment by ararslan** [comment]
Nanosoldier is broken right now anyway.

**Comment by andyferris** [comment]
OK, I tend to agree with Jameson, such changes should be at the very least be considered separately (I had originally figured the less independent "magic" parameters, the better).

This PR now merely tweaks the default inference parameters and leaves `Any16` alone.

**Comment by andyferris** [comment]
Bump

**Comment by martinholters** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by vtjnash** [comment]
Fixes https://github.com/JuliaLang/julia/issues/22370

In that issue, I recommended removing `limit_tuple_type`, so this seems to be a cautious start in that direction.

**Comment by ChrisRackauckas** [comment]
I'm not sure it's a full fix for #22370 which asks for a local override. While raising the global max helps, one could still run into a case where you know you have an array of 40 dimensions and just want a macro to locally change the max.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27398#issuecomment-397274461) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/86ce9626588ddd77c3de8327bcfdfce4faf574ae/354e1cf_vs_bad2a53/report.md). cc @ararslan

**Comment by andyferris** [comment]
Is that a typical amount of noise for nanosoldier? It seems unlikely to me that this change should affect many of the benchmarks...

**Comment by andyferris** [comment]
> In that issue, I recommended removing `limit_tuple_type`, so this seems to be a cautious start in that direction.

I'd also be happy if we simply removed the tuple limitations from inference, so long as this won't cause problems (but that shouldn't mean we can't merge this for now).

**Comment by ChrisRackauckas** [comment]
Bumping this. For reference, I received a PR which uses length 46 SVectors, and it's not a particularly bad idea in this case either.

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/374/files/581bf0e28fac0f5490cfc5bdc484ad531546de2c..9833eb62f4c2b4b6eda6cca62182428611bd33d6#diff-6c62ad55534b8f93c40614616894d756R2

**Comment by andyferris** [comment]
@ChrisRackauckas Yes, I believe (hope) we will be getting something like this in soon. After brief discussions with Jameson, my understanding is that we might try to eliminate the `tupletype_len` parameter altogether, which I planned to attempt on this branch when I get a moment.

Out of curiousity, did you experience performance problems working with `SVector{46, Int}`?

**Comment by KristofferC** [comment]
Isn't this a cache that is not supposed to be created over and over in e.g. a hot loop? Creating a length 46 array take on the order for 50ns so why are you worried about the allocation? Did you profile to see that using an Array slows things down? It might be faster?

**Comment by ChrisRackauckas** [comment]
>Isn't this a cache that is not supposed to be created over and over in e.g. a hot loop? Creating a length 46 array take on the order for 50ns so why are you worried about the allocation? Did you profile to see that using an Array slows things down? It might be faster?

It has nothing to do with that. Performance on the CPU with standard types etc. is probably fine with an array. But there are many reasons to want to write an algorithm with a loop-able list of more than 16 constants. In this particular application it has to do with attempts to compile to alternative targets like asm.js and .ptx kernels. Getting that stuff working is much easier when there are no arrays involved (and even if we can get it to work, in that case not having to create 4000 small GPU-based arrays to solve each small ODE in parallel is probably a good idea anyways)

**Comment by andyferris** [comment]
@vtjnash I've made a first attempt at erasing `tupletype_len` from existence, in the second commit. Is this now something like you imagined?

If not, we can try sneak just the first commit in v0.7 ASAP.

**Comment by andyferris** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/27398#issuecomment-400289787):
```
ProcessExitedException()
```
cc @ararslan

**Comment by ararslan** [comment]
Sorry, had to kill the job because the server needed to be restarted.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by andyferris** [comment]
CircleCI seems unique in complaining that 

```
From worker 3:	Internal error: encountered unexpected error in runtime:
      From worker 3:	ErrorException("type Params has no field MAX_TUPLETYPE_LEN")
```
Not sure how I can not be seeing this locally or in other CI? I've restarted of them...

**Comment by vtjnash** [comment]
Yes, this looks like what I had in mind

**Comment by andyferris** [comment]
OK - CircleCI was correctly picking up that @martinholters very recently merged some changes that depended on `MAX_TUPLETYPE_LEN` in #27434. @martinholters can you please check if the logic remains sound now I've removed that bound, or if we need to use the splat size parameter (`MAX_TUPLE_SPLAT`) to limit the loop [here](https://github.com/JuliaLang/julia/pull/27398/files#diff-493d285f483504268eb7cf24ea13b9aaR390)? Thanks.

**Comment by andyferris** [comment]
To follow up, I reinstated the limit in `abstract_iteration` to use `MAX_TUPLE_SPLAT` because this seems correct. I'm guessing that this type of iteration is used to infer e.g. `(key, value) = pair`, and so-on.

**Comment by andyferris** [comment]
@nanosoldier runbenchmarks(ALL, vs=":master")

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by martinholters** [comment]
Yes, using `MAX_TUPLE_SPLAT` there sounds correct :+1:

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27398#issuecomment-400405521) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ce1b4decd929fcf8ac0fb0438fdc3ae8fdd60d20/2522e21_vs_d8550d7/report.md). cc @ararslan

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -388,7 +387,7 @@ function abstract_iteration(@nospecialize(itertype), vtypes::VarTable, sv::Infer
     valtype = statetype = Bottom
     ret = Any[]
     stateordonet = widenconst(stateordonet)
-    while !(Nothing <: stateordonet) && length(ret) < sv.params.MAX_TUPLETYPE_LEN
+    while !(Nothing <: stateordonet) || length(ret) < sv.params.MAX_TUPLE_SPLAT
```
Comment (martinholters) [inline base/compiler/abstractinterpretation.jl]: Why the change from `&&` to `||`? Doesn't that lead to an infinite loop for something like
```julia
struct Foo end
Base.iterate(::Foo) = (0, nothing)
Base.iterate(::Foo, state) = (0, (state,))
foo(x) = tuple(x...)
@code_warntype foo(Foo())
```
?

Comment (andyferris) [reply base/compiler/abstractinterpretation.jl]: Oops!!! :anguished: I'll fix that

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27398#issuecomment-400553670) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/431daf78e65b03741f7db0035166613e23e4a7ff/6be931d_vs_d0e0e23/report.md). cc @ararslan

**Comment by JeffBezanson** [comment]
This is not release-blocking.

**Comment by ChrisRackauckas** [comment]
Why does the splat limit need to stay?

**Comment by andyferris** [comment]
I believe that it protects us from infinite recursion from badly written code, and the compiler never terminating.

I think this is because deciding whether a given piece of code is badly written is the same as the halting problem, which we can't solve. So we need some variety of simplistic, strong heuristic that guarantees compilation would not fail.

**Comment by GiggleLiu** [comment]
For tuple size > 16, the spatting still allocates. Is this an regression?

FYI: It causes problem to CUDA array manipulations, see issue
https://github.com/JuliaGPU/GPUArrays.jl/issues/340#issuecomment-745262184
